### ISSUE TITLE
Switch control flow convenience Spec type

### ISSUE THREAD

**Issue by kurtbrose** [mahmoud/glom#159]
it tries keys, and returns the value on a non-glom-error-raising key

```python
Switch({int: T, str: int})
```

To nest, Switch would need to be used twice.

For example, absolute value via nested Switch:

```python
Switch({
   int: Switch({
      M > 0: T,
      M < 0: -T,
   }),
})
```

**Comment by kurtbrose** [comment]
```python

class Switch(object):
   def __init__(self, cases):
      glom.match(cases, Or([(object, object], dict))
      self.cases = cases
   def glomit(self, target, scope):
      scope[MODE] = MATCH_MODE
      for match_key, fill_val in self.cases.items():
         try:
             scope[glom](target, match_key, scope)
             matched_fill_val = fill_val
         except GlomError:
             pass
      scope[MODE] = FILL_MODE
      try:
         return scope[glom](target, matched_fill_val, scope)
      except NameError:
         raise GlomError("no matches for target in Switch")

```

**Comment by kurtbrose** [comment]
```python

Ref('json', Switch({
    dict: {Ref('json'): Ref('json')},
    list: Auto([Ref('json')]),
    type(u''): str,
    object: T
}))
```

**Comment by mahmoud** [comment]
pseudocode looks pretty in line w what we discussed. do we want to take a default kwarg or at least establish a convention for what the wildcard should be?

**Comment by mahmoud** [comment]
`T: T` or `object: object`, maybe? Ideally wouldn't need to import another thing, so I lean a little toward object. Assuming we want a wildcard entry in the dict.

**Comment by mahmoud** [comment]
```python
abs_val = Switch({M < 0: -T, T: T})  # what does M do if it's the wrong type?
abs_val = Switch([(M < 0, T), (int, T)])  # get an error if it's not an int
```

**Comment by kurtbrose** [comment]
So, default behavior:

* target and results are ignored (just along-for-the-ride)
* GlomError counts as False
* other errors are let through
* on pass, target is returned; on failure, error is raised (vs return True/False)

All of these behaviors could be modified.

The whole list of possible arguments I can think of:

* list of passing values
* list of failing values
* function that returns True/False from a value
* list of falsey exception types
* function that determines if exception should be falsey or re-raised

In addition:

* "default" which returns if the bool would otherwise fail

At a high level, I think the simplest control would just be "bool-mode" or "glom-mode".

In bool mode, no exceptions are caught; True/False are returned.

In glom mode, target/result are ignored; result is passed through or GlomError raised.

**Comment by kurtbrose** [comment]
We have examples in schema, Check, and remap of a mix of exception and falsey behavior at the same time.

The challenge here is that control flow should be chainable.  That is, it is probably not coherent to have And/Or operate on falsiness of their inputs, and then raise GlomError on failure.  (If nothing else, they currently do not original errors.)  Conversely, we probably don't want a mode that ignores falsiness of input but returns True/False because it does not chain.

That leaves:

* "glom mode" -- exception control flow as in current behavior
* "bool mode" -- boolean logic as in python
* "bool mode with exception catching" -- truthiness of input and GlomErrors are used

**Comment by kurtbrose** [comment]
Other than default, this seems like the kind of thing we might want to circle back around to in the future rather than go too deep on for now.

**Comment by kurtbrose** [comment]
https://github.com/mahmoud/glom/pull/163

**Comment by mahmoud** [comment]
Fixed in #163: https://glom.readthedocs.io/en/latest/matching.html#control-flow-with-switch
